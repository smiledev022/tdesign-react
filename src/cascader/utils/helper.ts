import { TreeNode } from '../interface';

/**
 * 获取node节点的全路径
 * @param node
 * @returns
 */
export function getFullPathLabel(node: TreeNode) {
  return node
    .getPath()
    .map((node: TreeNode) => node.label)
    .join('/');
}
