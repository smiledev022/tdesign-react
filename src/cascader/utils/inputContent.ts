import { TreeNode, CascaderContextType, TreeNodeValue } from '../interface';

/**
 * icon Class
 * @param prefix
 * @param CLASSNAMES
 * @param cascaderContext
 * @returns
 */
export function getIconClass(prefix: string, CLASSNAMES: any, cascaderContext: CascaderContextType) {
  const { visible } = cascaderContext;
  return [
    `${prefix}-cascader-icon`,
    {
      [CLASSNAMES.STATUS.visible]: visible,
    },
  ];
}

/**
 * Cascader外显div class
 * @param prefix
 * @param CLASSNAMES
 * @param cascaderContext
 * @returns
 */
export function getCascaderInnerClasses(prefix: string, CLASSNAMES: any, cascaderContext: CascaderContextType) {
  const { disabled, visible, size, multiple } = cascaderContext;
  return [
    `${prefix}-cascader`,
    {
      [CLASSNAMES.STATUS.disabled]: disabled,
      [CLASSNAMES.STATUS.active]: visible,
      [CLASSNAMES.SIZE[size]]: size,
      [`${name}-is-multiple`]: multiple,
    },
  ];
}

/**
 * closeIcon 是否显示
 * @param isHover
 * @param cascaderContext
 * @returns
 */
export function getCloseShow(isHover: boolean, cascaderContext: CascaderContextType) {
  const { multiple, model, disabled, clearable, visible } = cascaderContext;
  return !!(
    !visible &&
    clearable &&
    isHover &&
    !disabled &&
    ((!multiple && model) || (multiple && (model as TreeNodeValue[]).length))
  );
}

/**
 * Placeholder 是否显示
 * @param isHover
 * @param cascaderContext
 * @returns
 */
export function getPlaceholderShow(
  cascaderContext: CascaderContextType,
  singleContent: string,
  multipleContent: TreeNode[],
) {
  const { filterable, visible, multiple } = cascaderContext;
  const valEmpty = !(multiple ? multipleContent.length : singleContent);
  if (filterable) {
    return valEmpty && !visible;
  }
  return valEmpty;
}

/**
 * 单选状态下内容
 * @param isHover
 * @param cascaderContext
 * @returns
 */
export function getSingleContent(cascaderContext: CascaderContextType) {
  const { model, multiple, treeStore, showAllLevels } = cascaderContext;
  if (multiple || !model) return;
  const node = treeStore && treeStore.getNodes(model as TreeNodeValue | TreeNode);
  if (!(node && node.length)) {
    return '';
  }
  const path = node && node[0].getPath();
  if (path && path.length) {
    return showAllLevels ? path.map((node: TreeNode) => node.label).join('/') : path[path.length - 1].label;
  }
  return model as string;
}

/**
 * 多选状态下选中内容
 * @param isHover
 * @param cascaderContext
 * @returns
 */
export function getMultipleContent(cascaderContext: CascaderContextType) {
  const { model, multiple, treeStore } = cascaderContext;

  if (!multiple || !model) return [];
  const node = treeStore && treeStore.getNodes(model as TreeNodeValue | TreeNode);
  if (!node) return [];
  const path = (model as TreeNodeValue[]).map((item: TreeNodeValue) => {
    const node = treeStore.getNodes(item);
    return node[0];
  });
  if (model && (model as TreeNodeValue[]).length) {
    return path && path.length ? path : (model as TreeNode[]);
  }
  return [];
}

/**
 * 点击popup outside副作用
 * @param ref
 * @param cascaderContext
 * @param event
 * @returns
 */
export function outerClickListenerEffect(
  ref: HTMLElement,
  cascaderContext: CascaderContextType,
  event: MouseEvent | TouchEvent,
) {
  const { visible, setVisible } = cascaderContext;
  if (!ref || ref.contains(event.target as Node)) {
    return;
  }

  if (visible) {
    setVisible(false);
  }
}

/**
 * closeIcon点击副作用
 * @param cascaderContext
 */
export function closeIconClickEffect(cascaderContext: CascaderContextType) {
  const { setVisible, setModel, treeStore, multiple } = cascaderContext;

  setModel(multiple ? [] : '');
  treeStore.resetChecked();
  treeStore.resetExpanded();
  setVisible(false);
}

/**
 * tag 关闭按钮点击副作用
 * @param cascaderContext
 */
export function handleRemoveTagEffect(cascaderContext: CascaderContextType, node: TreeNode) {
  const { disabled, setModel } = cascaderContext;

  if (disabled) return;
  const checked = node.setChecked(!node.isChecked());
  setModel(checked);
}

/**
 * 外显content点击副作用
 * @param cascaderContext
 */
export function innerContentClickEffect(cascaderContext: CascaderContextType) {
  const { setVisible, visible, filterActive, disabled } = cascaderContext;

  if (!disabled) {
    if (visible && filterActive) return; // input filter is active
    setVisible(!visible);
  }
}
