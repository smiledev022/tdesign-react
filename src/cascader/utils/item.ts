import { CascaderContextType, TreeNode } from '../interface';

/**
 * 是否显示省略计算方法
 * @param node
 * @param size
 * @returns
 */
export function getLabelIsEllipsis(node: TreeNode, size: CascaderContextType['size']) {
  const sizeMap = {
    small: 11,
    medium: 9,
    large: 8,
  };
  return sizeMap[size] < node.label.length;
}

export function getCascaderItemClass(
  prefix: string,
  node: TreeNode,
  CLASSNAMES: any,
  cascaderContext: CascaderContextType,
) {
  const { checkStrictly, multiple, size } = cascaderContext;
  const expandedActive = !checkStrictly && node.expanded && (multiple ? !node.isLeaf() : true);

  return [
    `${prefix}-cascader-item`,
    {
      [CLASSNAMES.STATUS.selected]: node.checked,
      [CLASSNAMES.STATUS.expanded]: expandedActive,
      [CLASSNAMES.STATUS.disabled]: node.disabled,
      [CLASSNAMES.STATUS.active]: node.actived || (multiple && node.expanded),
      [CLASSNAMES.SIZE[size]]: size,
      [`${prefix}-cascader-item-have-icon`]: node.children,
    },
  ];
}

export function getCascaderItemIconClass(
  prefix: string,
  node: TreeNode,
  CLASSNAMES: any,
  cascaderContext: CascaderContextType,
) {
  const { checkStrictly } = cascaderContext;

  return [
    `${prefix}-cascader-icon`,
    {
      [CLASSNAMES.STATUS.expanded]: !checkStrictly && node.expanded,
    },
  ];
}
