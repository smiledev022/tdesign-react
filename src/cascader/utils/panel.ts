import isFunction from 'lodash/isFunction';
import isNumber from 'lodash/isNumber';
import { TreeNode, CascaderContextType, CascaderProps, ContextType } from '../interface';

/**
 * 面板数据计算方法
 * @param treeNodes
 * @returns
 */
export function getPanels(treeNodes: CascaderContextType['treeNodes']) {
  const panels: TreeNode[][] = [];
  treeNodes.forEach((node: TreeNode) => {
    if (panels[node.level]) {
      panels[node.level].push(node);
    } else {
      panels[node.level] = [node];
    }
  });
  return panels;
}

/**
 * 点击item的副作用
 * @param propsTrigger
 * @param trigger
 * @param node
 * @param cascaderContext
 * @param onChange
 * @param ctx
 */
export function expendClickEffect(
  propsTrigger: CascaderProps['trigger'],
  trigger: CascaderProps['trigger'],
  node: TreeNode,
  cascaderContext: CascaderContextType,
  onChange: CascaderProps['onChange'],
  ctx: any,
) {
  const {
    checkStrictly,
    filterActive,
    multiple,
    treeStore,
    setFilterActive,
    setVisible,
    setModel,
    setTreeNodes,
    setExpend,
  } = cascaderContext;

  // 点击展开节点，设置展开状态
  if (propsTrigger === trigger && !node.isLeaf()) {
    const expanded = node.setExpanded(true);
    treeStore.refreshNodes();
    treeStore.replaceExpanded(expanded);
    const nodes = treeStore.getNodes().filter((node: TreeNode) => node.visible);
    setTreeNodes(nodes);

    // 多选条件下手动维护expend
    if (multiple) {
      setExpend(expanded);
    }
  }

  if (!multiple && (node.isLeaf() || checkStrictly) && trigger === 'click') {
    treeStore.resetChecked();
    const checked = node.setChecked(!node.isChecked());
    const [value] = checked;

    if (onChange && isFunction(onChange)) {
      onChange(value, ctx);
    }

    // 过滤状态下，点击后清除过滤状态
    if (filterActive) {
      setFilterActive(false);
    }

    // 非过滤状态下，关闭
    if (!filterActive) {
      setVisible(false);
    }

    // 非受控状态下更新状态
    setModel(value);
  }
}

/**
 * 多选状态下选中状态数据变化的副作用
 * @param node
 * @param cascaderContext
 * @param onChange
 * @param ctx
 * @returns
 */
export function valueChangeEffect(
  node: TreeNode,
  cascaderContext: CascaderContextType,
  onChange: CascaderProps['onChange'],
  ctx: ContextType,
) {
  const { disabled, max, multiple, setVisible, setModel, filterActive, setFilterActive, treeNodes, treeStore } =
    cascaderContext;

  if (!node || disabled || node.disabled) {
    return;
  }
  const checked = node.setChecked(!node.isChecked());

  if (isNumber(max) && max < 0) {
    console.warn('max should > 0');
  }

  if (checked.length > max && isNumber(max) && max > 0) {
    return;
  }

  if (checked.length === 0) {
    const expanded = treeStore.getExpanded();
    setTimeout(() => {
      treeStore.replaceExpanded(expanded);
      treeStore.refreshNodes();
    }, 0);
  }

  if (!multiple) {
    setVisible(false);
  }

  const isSelectAll = treeNodes.every((item) => checked.indexOf(item.value) > -1);

  if (filterActive && isSelectAll) {
    setVisible(false);
    setFilterActive(false);
  }

  setModel(checked);

  if (onChange && isFunction(onChange)) {
    onChange(checked, ctx);
  }
}
