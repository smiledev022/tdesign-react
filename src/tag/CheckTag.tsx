import React, { forwardRef } from 'react';
import classNames from 'classnames';
import { StyledProps, Combine, Omit } from '../_type';
import useConfig from '../_util/useConfig';

/**
 * CheckTag 组件支持的属性。
 *
 */
export interface CheckTagProps
  extends Combine<
    StyledProps,
    Omit<React.ButtonHTMLAttributes<HTMLButtonElement>, 'type'>
  > {
  /**
   * 设置按钮为禁用状态
   *
   *@default false
   * */
  disabled?: boolean;
  /**
   * 选中状态
   *
   * @default false
   */
  checked?: boolean;
}

export const CheckTag = forwardRef(
  (props: CheckTagProps, ref: React.Ref<HTMLButtonElement>) => {
    const {
      checked,
      disabled,
      children,
      style,
      className,
      ...tagOtherProps
    } = props;

    // 前缀声明
    const { classPrefix } = useConfig();
    const tagClassPrefix = `${classPrefix}-tag`;

    // 属性判断
    const classList = {
      [`${tagClassPrefix}--disabled`]: disabled,
      [`${tagClassPrefix}--checked`]: checked,
    };

    // 添加默认属性
    const checkTagClassNames = classNames(
      tagClassPrefix,
      `${tagClassPrefix}--default`,
      classList
    );

    // 合并用户自定义className
    const mergedClassName = `${checkTagClassNames} ${className || ''}`;

    const checkTag: JSX.Element = (
      <span
        ref={ref}
        className={mergedClassName}
        style={style}
        {...tagOtherProps}
      >
        {children}
      </span>
    );

    return checkTag;
  }
);

CheckTag.displayName = 'TDesingCheckTag';
