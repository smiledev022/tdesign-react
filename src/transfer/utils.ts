import isEmpty from 'lodash/isEmpty';
import isFunction from 'lodash/isFunction';
import isString from 'lodash/isString';
import React from 'react';
import { TNode } from '../common';

export const getDefaultValue = <T>(value: T | T[], defaultValue?: T): T[] => {
  if (isEmpty(value)) {
    return Array.isArray(defaultValue) ? defaultValue : [defaultValue, defaultValue];
  }

  if (Array.isArray(value)) return value;

  return [defaultValue, defaultValue];
};

export const getJSX = (value: string | TNode) => {
  if (isString(value)) return value;

  if (isFunction(value)) return value();

  if (React.isValidElement(value)) return value;

  return value;
};
