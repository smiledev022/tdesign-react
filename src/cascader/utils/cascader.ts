import { getFullPathLabel } from './helper';
import { TreeNode, TreeNodeValue, TreeOptionData, CascaderContextType } from '../interface';

/**
 * treeValue计算方法
 * @param model
 * @returns
 */
export const getTreeValue = (model: CascaderContextType['model']) => {
  let treeValue: TreeNodeValue[] = [];
  if (Array.isArray(model)) {
    if (model.length > 0 && typeof model[0] === 'object') {
      treeValue = (model as TreeOptionData[]).map((val) => val.value);
    } else if (model.length) {
      treeValue = model as TreeNodeValue[];
    }
  } else if (model) {
    if (typeof model === 'object') {
      treeValue = [(model as TreeOptionData).value];
    } else {
      treeValue = [model];
    }
  }
  return treeValue;
};

/**
 * input和treeStore变化的副作用
 * @param inputVal
 * @param treeStore
 * @param setTreeNodes
 * @returns
 */
export const treeNodesEffect = (
  inputVal: CascaderContextType['inputVal'],
  treeStore: CascaderContextType['treeStore'],
  setTreeNodes: CascaderContextType['setTreeNodes'],
) => {
  if (!treeStore) return;
  let nodes = [];
  if (inputVal) {
    nodes = treeStore.nodes.filter((node: TreeNode) => {
      const fullPathLabel = getFullPathLabel(node);
      return fullPathLabel.indexOf(inputVal) > -1 && node.isLeaf();
    });
  } else {
    nodes = treeStore.getNodes().filter((node: TreeNode) => node.visible);
  }
  setTreeNodes(nodes);
};

/**
 * 初始化展开阶段与展开状态副作用
 * @param treeStore
 * @param treeValue
 * @param expend
 */
export const treeStoreExpendEffect = (
  treeStore: CascaderContextType['treeStore'],
  treeValue: TreeNodeValue[],
  expend: TreeNodeValue[],
) => {
  // init expanded, 无expend状态时设置
  if (Array.isArray(treeValue) && expend.length === 0) {
    const expandedMap = new Map();
    const [val] = treeValue;
    if (val) {
      expandedMap.set(val, true);
      const node = treeStore.getNode(val);
      if (!node) return;
      node.getParents().forEach((tn: TreeNode) => {
        expandedMap.set(tn.value, true);
      });
      const expandedArr = Array.from(expandedMap.keys());
      treeStore.replaceExpanded(expandedArr);
    }
  }
  // 本地维护 expend，更加可控，不需要依赖于 tree 的状态
  if (treeStore.getExpanded() && expend.length) {
    treeStore.replaceExpanded(expend);
  }
  treeStore.refreshNodes();
};
