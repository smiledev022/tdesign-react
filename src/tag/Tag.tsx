import React, { forwardRef } from 'react';
import classNames from 'classnames';
import { StyledProps, Combine, Omit } from '../_type';
import { CheckTag, CheckTagProps } from './CheckTag';
import useConfig from '../_util/useConfig';

/**
 * Tag 组件支持的属性。
 */
export interface TagProps
  extends Combine<StyledProps, Omit<React.ButtonHTMLAttributes<HTMLButtonElement>, 'type'>> {
  /**
   * 标签类型
  /*
   * theme?: 'default'|'primary'|'info'|'warning'|'danger'|'success'
   *
   *  @default "middle"
   */
  theme?: string;
  /**
   * 按钮大小
   *
   * @default "middle"
   */
  size?: 'large' | 'middle' | 'small';
  /**
   * 设置按钮为禁用状态
   *
   * @default false
   * */
  disabled?: boolean;
  /**
   * 设置按钮为禁用状态
   *
   *@default false
   * */
  closable?: boolean;
  /**
   * 最大长度
   *
   */
  maxWidth?: string | number;
  /**
   * 形状
   *
   * @default 'square'
   */
  shape?: 'square' | 'round' | 'mark';
  /**
   * 样式模式
   *
   * @default 'dark'
   */
  effect?: 'dark' | 'light' | 'plain';
  /**
   * 点击回调函数
   */
  onClick?: (e?: React.MouseEvent) => void;

  /**
   * 关闭回调函数
   */
  onClose?: (e?: React.MouseEvent) => void;
}

/*
 * 因为checkTag不属于props，需要单独写一个声明
 * 参照 https://segmentfault.com/q/1010000019012848/
 */

interface ExtraTagRef extends React.ForwardRefExoticComponent<TagProps> {
  CheckTag: React.FunctionComponent<CheckTagProps>;
}

/**
 * 标签组件
 *
 * @see https://tdesign.tencent.com/react/tag
 */
export const Tag = forwardRef((props: TagProps, ref: React.Ref<HTMLButtonElement>) => {
  const {
    theme = 'default',
    size = 'middle',
    effect = 'dark',
    closable,
    maxWidth,
    shape = 'square',
    disabled,
    className,
    style,
    onClick,
    onClose,
    children,
    ...otherTagProps
  } = props;

  const { classPrefix } = useConfig();
  const tagClassPrefix = `${classPrefix}-tag`;

  /**
   * 根据用户配置,获取标签class列表
   */
  const classList = {
    [`${tagClassPrefix}--ellipsis`]: !!maxWidth,
    [`${tagClassPrefix}--disabled`]: disabled,
  };
  /**
   * size比较特殊，用的是简写
   */
  const sizeMap = {
    large: `${classPrefix}-size-l`,
    small: `${classPrefix}-size-s`,
  };

  /*
   * 解析所有参数，获取对应的className
   */
  const tagClassNames = classNames(
    tagClassPrefix,
    `${tagClassPrefix}--${theme}`,
    `${tagClassPrefix}--${effect}`,
    `${tagClassPrefix}--${size}`,
    `${tagClassPrefix}--${shape}`,
    classList,
    sizeMap[size] || '',
  );

  /**
   * 校验maxwidth
   *
   * @param {string} classType class类型
   * @param {string}  classVal 用户传过来的class值
   * @return {string} 返回class名称
   */
  const getMaxWidth = (maxWidth: string | number): string | number => {
    // 没有%并且没有px的时候，手动添加%
    const isNeedPx = `${maxWidth}`.indexOf('px') === -1 && `${maxWidth}`.indexOf('%') === -1;

    return isNeedPx ? `${maxWidth}px` : maxWidth;
  };

  /**
   * 处理用户提供的最大长度，转化为标准的css
   */
  const formatMaxWidth = getMaxWidth(maxWidth);
  const tagStyle = { ...style, ...{ maxWidth: formatMaxWidth } };

  /**
   * 删除icon变量
   */
  const deleteIcon = <i className="t-icon t-icon-close" onClick={onClose || null}></i>;

  /**
   * 合并用户自己传过来长度className
   */
  const mergedClassName = `${tagClassNames} ${className || ''}`;

  const tag: JSX.Element = (
    <span
      ref={ref}
      className={mergedClassName}
      onClick={onClick || null}
      style={tagStyle}
      {...otherTagProps}
    >
      {children}
      {closable && deleteIcon}
    </span>
  );

  return tag;
}) as ExtraTagRef;

/*
 * 给Tag新增一个CheckTag组件
 */
Tag.CheckTag = CheckTag;

Tag.displayName = 'Tag';
