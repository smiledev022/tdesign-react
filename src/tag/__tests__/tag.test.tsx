import React from 'react';
import { testExamples, render, getByText, fireEvent } from '@test/utils';
import Tag from '../Tag';
import CheckTag from '../CheckTag';
import ClosableTag from '../_example/ClosableTag';
import DisableTag from '../_example/DisableTag';

// 测试组件代码 Example 快照
testExamples(__dirname);

describe('Tag 组件测试', () => {
  test('tag 默认的样式和class', async () => {
    const tagTitle = '默认标签';
    const { container } = render(<Tag>{tagTitle}</Tag>);

    // span标签 内容正常显示
    getByText(container, (_, element) => element.tagName.toLowerCase() === 'span');

    // 校验默认className
    const defaultClass = ['t-tag', 't-tag--dark', 't-tag--middle', 't-tag--square'];
    expect(container.firstChild).toHaveClass(...defaultClass);
  });

  test('ClosableTag 点击后关闭按钮', async () => {
    const tagRegExp = /点击关闭/;

    const { queryAllByText, getByText } = render(<ClosableTag></ClosableTag>);
    // 点击i标签后，关闭一个，3个变2个
    expect(queryAllByText(tagRegExp).length).toEqual(3);
    fireEvent.click(getByText('点击关闭0').querySelector('i'));
    expect(queryAllByText(tagRegExp).length).toEqual(2);
  });

  test('Tag  文字超长截断', async () => {
    const tagLabel = '超长省略文本标签超长省略文本标签';

    const { container } = render(<Tag maxWidth={150}>{tagLabel}</Tag>);
    // 最长宽度为150
    expect(container.firstChild).toHaveStyle({ 'max-width': '150px' });
  });
});

describe('CheckTag 组件测试', () => {
  test('CheckTag 选中后变色', async () => {
    const testTagText = '标签1';
    const checkedClass = 't-tag--checked';

    const { getByText, container } = render(<CheckTag>{testTagText}</CheckTag>);
    // 点击后变色
    expect(container.firstChild).not.toHaveClass(checkedClass);
    fireEvent.click(getByText(testTagText));
    expect(container.firstChild).toHaveClass(checkedClass);
  });

  test('Tag 无效时不能点击', async () => {
    const { getByText } = render(<DisableTag />);

    fireEvent.click(getByText('有效标签'));
    fireEvent.click(getByText('无效标签'));

    expect(getByText('有效已选中')).not.toBeNull();
  });
});
