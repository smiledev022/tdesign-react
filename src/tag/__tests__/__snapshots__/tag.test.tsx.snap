// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`base.jsx 1`] = `
<DocumentFragment>
  <div
    style="width: 285px; display: flex; justify-content: space-around;"
  >
    <span>
      默认：
    </span>
    <span
      class="t-tag t-tag--default t-tag--dark t-tag--middle t-tag--square"
    >
      标签一
    </span>
    <a
      href="https://www.tencent.com/zh-cn"
      rel="noreferrer"
      target="_blank"
    >
      <span
        class="t-tag t-tag--default t-tag--dark t-tag--middle t-tag--square"
      >
        超链接
      </span>
    </a>
  </div>
  <div
    style="width: 500px; display: flex; justify-content: space-around;"
  >
    <span>
      深色：
    </span>
    <span
      class="t-tag t-tag--primary t-tag--dark t-tag--middle t-tag--square"
    >
      标签一
    </span>
    <span
      class="t-tag t-tag--warning t-tag--dark t-tag--middle t-tag--square"
    >
      标签二
    </span>
    <span
      class="t-tag t-tag--danger t-tag--dark t-tag--middle t-tag--square"
    >
      标签三
    </span>
    <span
      class="t-tag t-tag--success t-tag--dark t-tag--middle t-tag--square"
    >
      标签四
    </span>
  </div>
  <div
    style="width: 500px; display: flex; justify-content: space-around;"
  >
    <span>
      浅色：
    </span>
    <span
      class="t-tag t-tag--primary t-tag--light t-tag--middle t-tag--square"
    >
      标签一
    </span>
    <span
      class="t-tag t-tag--warning t-tag--light t-tag--middle t-tag--square"
    >
      标签二
    </span>
    <span
      class="t-tag t-tag--danger t-tag--light t-tag--middle t-tag--square"
    >
      标签三
    </span>
    <span
      class="t-tag t-tag--success t-tag--light t-tag--middle t-tag--square"
    >
      标签四
    </span>
  </div>
  <div
    style="width: 500px; display: flex; justify-content: space-around;"
  >
    <span>
      朴素：
    </span>
    <span
      class="t-tag t-tag--primary t-tag--plain t-tag--middle t-tag--square"
    >
      标签一
    </span>
    <span
      class="t-tag t-tag--warning t-tag--plain t-tag--middle t-tag--square"
    >
      标签二
    </span>
    <span
      class="t-tag t-tag--danger t-tag--plain t-tag--middle t-tag--square"
    >
      标签三
    </span>
    <span
      class="t-tag t-tag--success t-tag--plain t-tag--middle t-tag--square"
    >
      标签四
    </span>
  </div>
</DocumentFragment>
`;

exports[`delete.jsx 1`] = `
<DocumentFragment>
  <div
    style="display: flex;"
  >
    <span
      style="margin-right: 30px;"
    >
      可删除：
    </span>
    <span
      class="t-tag t-tag--default t-tag--dark t-tag--middle t-tag--square"
      style="margin-right: 30px;"
    >
      可删除标签0
      <svg
        class="t-icon t-icon-close"
        version="1.1"
        viewBox="0 0 1024 1024"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M512.0032 570.83072L708.864 767.6864 767.6928 708.8576 570.83456 511.99936 767.6928 315.14176 708.864 256.3104 512.0032 453.168 315.14368 256.30848 256.31232 315.13984 453.17184 511.99936 256.31232 708.8576 315.14368 767.6928 512.0032 570.83072Z"
        />
      </svg>
    </span>
    <span
      class="t-tag t-tag--default t-tag--dark t-tag--middle t-tag--square"
      style="margin-right: 30px;"
    >
      <svg
        class="t-icon t-icon-discount"
        version="1.1"
        viewBox="0 0 1024 1024"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M493.04229568 116.1597152C508.86786752 116.30420928 524.0067104 121.97761344 535.39539904 133.36630208L942.6889024 540.659808C967.6824576 565.65336256 965.1497984 608.70858112 937.0320512 636.82633024L631.56192128 942.2964608C603.44417216 970.414208 560.3889536 972.9468672 535.39539904 947.953312L128.10189312 540.659808C116.179456 528.73737088 110.52056832 512.70502848 110.91453056 496.14468224L110.89530624 496.1568576 110.89530624 180.1597152C110.89530624 147.33822144 135.6018784 120.28725696 167.43155456 116.59028992L174.89530624 116.1597152 493.04229568 116.1597152ZM174.89530624 180.0957152L174.89530624 496.1568576 174.06330624 496.1277152 580.65023296 902.6984768 897.4340736 585.91464192 491.69530624 180.1597152 174.89530624 180.0957152ZM396.1453088 256.40588416C466.8377568 256.40588416 524.1453088 313.71343616 524.1453088 384.40588416 524.1453088 455.09833216 466.8377568 512.40588416 396.1453088 512.40588416 325.4528608 512.40588416 268.1453088 455.09833216 268.1453088 384.40588416 268.1453088 313.71343616 325.4528608 256.40588416 396.1453088 256.40588416ZM396.1453088 320.40588416C360.7990848 320.40588416 332.1453088 349.05966016 332.1453088 384.40588416 332.1453088 419.75210816 360.7990848 448.40588416 396.1453088 448.40588416 431.4915328 448.40588416 460.1453088 419.75210816 460.1453088 384.40588416 460.1453088 349.05966016 431.4915328 320.40588416 396.1453088 320.40588416Z"
          fill-opacity=".9"
          fill-rule="evenodd"
        />
      </svg>
      可删除标签1
      <svg
        class="t-icon t-icon-close"
        version="1.1"
        viewBox="0 0 1024 1024"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M512.0032 570.83072L708.864 767.6864 767.6928 708.8576 570.83456 511.99936 767.6928 315.14176 708.864 256.3104 512.0032 453.168 315.14368 256.30848 256.31232 315.13984 453.17184 511.99936 256.31232 708.8576 315.14368 767.6928 512.0032 570.83072Z"
        />
      </svg>
    </span>
    <span
      class="t-tag t-tag--default t-tag--dark t-tag--middle t-tag--square t-tag--disabled"
      style="margin-right: 30px;"
    >
      可删除标签2
      <svg
        class="t-icon t-icon-close"
        version="1.1"
        viewBox="0 0 1024 1024"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M512.0032 570.83072L708.864 767.6864 767.6928 708.8576 570.83456 511.99936 767.6928 315.14176 708.864 256.3104 512.0032 453.168 315.14368 256.30848 256.31232 315.13984 453.17184 511.99936 256.31232 708.8576 315.14368 767.6928 512.0032 570.83072Z"
        />
      </svg>
    </span>
  </div>
  <div
    style="display: flex; cursor: pointer;"
  >
    <span
      style="margin-right: 30px;"
    >
      可添加：
    </span>
    <span
      class="t-tag t-tag--default t-tag--dark t-tag--middle t-tag--square"
    >
      <svg
        class="t-icon t-icon-add"
        version="1.1"
        viewBox="0 0 1024 1024"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M470.39872 553.59872V800H553.59872V553.59872H800V470.39872H553.59872V224H470.39872V470.39872H224V553.59872H470.39872Z"
        />
      </svg>
      可添加标签
    </span>
  </div>
</DocumentFragment>
`;

exports[`icon.jsx 1`] = `
<DocumentFragment>
  <span
    class="t-tag t-tag--default t-tag--dark t-tag--middle t-tag--square"
  >
    <svg
      class="t-icon t-icon-discount"
      version="1.1"
      viewBox="0 0 1024 1024"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M493.04229568 116.1597152C508.86786752 116.30420928 524.0067104 121.97761344 535.39539904 133.36630208L942.6889024 540.659808C967.6824576 565.65336256 965.1497984 608.70858112 937.0320512 636.82633024L631.56192128 942.2964608C603.44417216 970.414208 560.3889536 972.9468672 535.39539904 947.953312L128.10189312 540.659808C116.179456 528.73737088 110.52056832 512.70502848 110.91453056 496.14468224L110.89530624 496.1568576 110.89530624 180.1597152C110.89530624 147.33822144 135.6018784 120.28725696 167.43155456 116.59028992L174.89530624 116.1597152 493.04229568 116.1597152ZM174.89530624 180.0957152L174.89530624 496.1568576 174.06330624 496.1277152 580.65023296 902.6984768 897.4340736 585.91464192 491.69530624 180.1597152 174.89530624 180.0957152ZM396.1453088 256.40588416C466.8377568 256.40588416 524.1453088 313.71343616 524.1453088 384.40588416 524.1453088 455.09833216 466.8377568 512.40588416 396.1453088 512.40588416 325.4528608 512.40588416 268.1453088 455.09833216 268.1453088 384.40588416 268.1453088 313.71343616 325.4528608 256.40588416 396.1453088 256.40588416ZM396.1453088 320.40588416C360.7990848 320.40588416 332.1453088 349.05966016 332.1453088 384.40588416 332.1453088 419.75210816 360.7990848 448.40588416 396.1453088 448.40588416 431.4915328 448.40588416 460.1453088 419.75210816 460.1453088 384.40588416 460.1453088 349.05966016 431.4915328 320.40588416 396.1453088 320.40588416Z"
        fill-opacity=".9"
        fill-rule="evenodd"
      />
    </svg>
    默认标签
  </span>
</DocumentFragment>
`;

exports[`long-text.jsx 1`] = `
<DocumentFragment>
  <span
    class="t-tag t-tag--default t-tag--dark t-tag--middle t-tag--square t-tag--ellipsis"
    style="margin-right: 5px; margin-bottom: 10px; max-width: 150px;"
  >
    超长省略文本标签超长省略文本标签
  </span>
</DocumentFragment>
`;

exports[`selectable.jsx 1`] = `
<DocumentFragment>
  <span
    class="t-tag t-tag--default t-tag--check t-tag--checked"
    style="margin-right: 30px;"
  >
    选中
  </span>
  <span
    class="t-tag t-tag--default t-tag--check"
    style="margin-right: 30px;"
  >
    未选
  </span>
  <span
    class="t-tag t-tag--default t-tag--check t-tag--disabled"
    style="margin-right: 30px;"
  >
    Disabled
  </span>
</DocumentFragment>
`;

exports[`shape.jsx 1`] = `
<DocumentFragment>
  <div
    style="width: 300px; display: flex; justify-content: space-around;"
  >
    <span>
      正常：
    </span>
    <span
      class="t-tag t-tag--default t-tag--dark t-tag--middle t-tag--square"
    >
      标签一
    </span>
    <span
      class="t-tag t-tag--primary t-tag--dark t-tag--middle t-tag--square"
    >
      标签一
    </span>
  </div>
  <div
    style="width: 300px; display: flex; justify-content: space-around;"
  >
    <span>
      圆角：
    </span>
    <span
      class="t-tag t-tag--default t-tag--dark t-tag--middle t-tag--round"
      style="margin-right: 5px;"
    >
      标签一
    </span>
    <span
      class="t-tag t-tag--primary t-tag--dark t-tag--middle t-tag--round"
    >
      标签一
    </span>
  </div>
  <div
    style="width: 300px; display: flex; justify-content: space-around;"
  >
    <span>
      标记：
    </span>
    <span
      class="t-tag t-tag--default t-tag--dark t-tag--middle t-tag--mark"
    >
      标签一
    </span>
    <span
      class="t-tag t-tag--primary t-tag--dark t-tag--middle t-tag--mark"
    >
      标签一
    </span>
  </div>
</DocumentFragment>
`;

exports[`size.jsx 1`] = `
<DocumentFragment>
  <span
    class="t-tag t-tag--default t-tag--dark t-tag--small t-tag--square t-size-s"
    style="margin-right: 30px;"
  >
    小型标签
  </span>
  <span
    class="t-tag t-tag--default t-tag--dark t-tag--medium t-tag--square"
    style="margin-right: 30px;"
  >
    默认标签
  </span>
  <span
    class="t-tag t-tag--default t-tag--dark t-tag--large t-tag--square t-size-l"
    style="margin-right: 30px;"
  >
    大型标签
  </span>
</DocumentFragment>
`;
